### PR TITLE
fix(router): ensure navigation via back button works

### LINKED ISSUE #2201: Router - back button does not work properly
Back button works well if only one route is navigated (renders the previous view). But if more than one route is followed, nothing happen when clicking the back button.
History is also not maintained, it is impossible to go forward.

### PR DESCRIPTION
**Pull request by matsko**
The router will now navigate and respect the current address value
accordingly whenever a popState event is handled.

Closes #2201

### CODE REVIEW

File: modules/angular2/test/router/router_integration_spec.ts [hunk 1]
```diff
@@ -81,6 +82,74 @@ export function main() {
...
+                 var counter = 0;
+
+                 router.subscribe((_) => {
+                   counter++;
+
+                   var location = rootTC.componentInstance.location;
+                   var element = rootTC.nativeElement;
+                   var path = location.path();
+
+                   /*
```
Comment (btford) [inline modules/angular2/test/router/router_integration_spec.ts]: is this something we want to assert happens in this order? maybe build up an array of strings representing these events and assert it looks like what we expect?

Comment (btford) [inline modules/angular2/test/router/router_integration_spec.ts]: Hmm, maybe that's not any better, because you want to also assert that the template is in the right state.

I'll leave it to your discretion.

File: modules/angular2/src/router/router.ts [hunk 2]
```diff
@@ -290,14 +290,17 @@ export class RootRouter extends Router {
               hostComponent: Type) {
     super(registry, pipeline, null, hostComponent);
     this._location = location;
-    this._location.subscribe((change) => this.navigate(change['url']));
+    this._location.subscribe((change) => this.navigate(change['url'], isPresent(change['pop'])));
```
Comment (btford) [inline modules/angular2/src/router/router.ts]: so if we emit a `change` like `{ path: 'foo', pop: false }`, we'll push a new state onto the history?

Comment (matsko) [inline modules/angular2/src/router/router.ts]: Yes or without pop at all. We do need a better name than pop though.

File: modules/angular2/test/router/router_integration_spec.ts [hunk 3]
```diff
@@ -81,6 +82,74 @@ export function main() {
          }));
     });
 
+    describe('back button app', () => {
+      beforeEachBindings(() => {
+        return [
```
Comment (btford) [inline modules/angular2/test/router/router_integration_spec.ts]: Why?

File: modules/angular2/test/router/router_spec.ts [hunk 4]
```diff
@@ -76,6 +76,20 @@ export function main() {
              });
        }));
 
+    it('should activate viewports but not update URL on navigate',
```
Comment (btford) [inline modules/angular2/test/router/router_spec.ts]: Seems not descriptive enough. Maybe `'should not push a history change on when navigate is called with skipUrlChange'` ?

File: modules/angular2/test/router/router_integration_spec.ts [hunk 5]
```diff
@@ -81,6 +82,74 @@ export function main() {
...
+          bind(appComponentTypeToken)
+              .toValue(HierarchyAppCmp),
+          bind(LocationStrategy).toClass(HashLocationStrategy)
+        ];
+      });
+
+      it('should navigate to a URL and back while properly changing back to the former URL',
```
Comment (btford) [inline modules/angular2/test/router/router_integration_spec.ts]: test descriptor seems kind of wordy– maybe something like "it should change the url without pushing a new history state for back navigations" ?

File: modules/angular2/test/router/router_integration_spec.ts (hunk 3 above)
Comment (matsko) [inline modules/angular2/test/router/router_integration_spec.ts]: This has been fixed now. There will be another commit that this one is on top of.

File: modules/angular2/test/router/router_integration_spec.ts (hunk 1 above)
Comment (matsko) [inline modules/angular2/test/router/router_integration_spec.ts]: Fixed. Please check next commit.

File: modules/angular2/test/router/router_spec.ts (hunk 4 above)
Comment (matsko) [inline modules/angular2/test/router/router_spec.ts]: Fixed.

File: modules/angular2/test/router/router_integration_spec.ts (hunk 5 above)
Comment (matsko) [inline modules/angular2/test/router/router_integration_spec.ts]: Fixed.

**Comment by matsko** [comment]
@btford there is a new commit now with a different strategy than that of using `switch`. Please review :)

**Comment by matsko** [comment]
:cat2:

**Comment by btford** [comment]
LGTM, merge. (remember to use new pre-submit hook)

File: modules/angular2/test/router/router_integration_spec.ts [hunk 6]
```diff
@@ -81,6 +82,57 @@ export function main() {
...
+               .then((rootTC) => {
+                 var router = rootTC.componentInstance.router;
+                 var position = 0;
+                 var flipped = false;
+                 var history =
+                     [
+                       ['/parent/child', 'root { parent { hello } }', '/super-parent/child'],
+                       ['/super-parent/child', 'root { super-parent { hello2 } }', '/parent/child'],
+                       ['/parent/child', 'root { parent { hello } }', false]
+                     ]
```
Comment (btford) [inline modules/angular2/test/router/router_integration_spec.ts]: missing `;`

Comment (matsko) [inline modules/angular2/test/router/router_integration_spec.ts]: fixed

**Comment by btford** [comment]
Landed as 7bf7ec6d9c8296a206864bbb7310cebe8e4d7259.
